internal/security/setup/option/cache_test.go

```diff
@@ -0,0 +1,209 @@
+	assert.False(deployEmpty)
+}
+
+func TestCacheOff(t *testing.T) {
+	vaultJSONPkiSetupExist = true
+	kongJSONPkiSetupExist = true
+	tearDown := setupCacheTest(t)
+	defer tearDown(t)
+
+	options := PkiInitOption{
+		CacheOpt: false,
+	}
+	cacheOff, _, _ := NewPkiInitOption(options)
+	cacheOff.(*PkiInitOption).executor = testExecutor
+	exitCode, err := cacheOff.executeOptions(Cache())
+
+	assert := assert.New(t)
+	assert.Equal(normal, exitCode)
+	assert.Nil(err)
+}
+
+func setupCacheTest(t *testing.T) func(t *testing.T) {
+	testExecutor = &mockOptionsExecutor{}
+	curDir, err := os.Getwd()
+	if err != nil {
+		t.Fatalf("cannot get the working dir %s: %v", curDir, err)
+	}
+
+	testResourceDir := filepath.Join(curDir, resourceDirName)
+	if err := createDirectoryIfNotExists(testResourceDir); err != nil {
+		t.Fatalf("cannot create resource dir %s for the test: %v", testResourceDir, err)
+	}
+
+	resDir := filepath.Join(curDir, "testdata", resourceDirName)
+	jsonVaultFile := filepath.Join(curDir, resourceDirName, pkiSetupVaultJSON)
+	if vaultJSONPkiSetupExist {
+		if _, err := copyFile(filepath.Join(resDir, pkiSetupVaultJSON), jsonVaultFile); err != nil {
+			t.Fatalf("cannot copy %s for the test: %v", pkiSetupVaultJSON, err)
+		}
+	}
+
+	jsonKongFile := filepath.Join(curDir, resourceDirName, pkiSetupKongJSON)
+	if kongJSONPkiSetupExist {
+		if _, err := copyFile(filepath.Join(resDir, pkiSetupKongJSON), jsonKongFile); err != nil {
+			t.Fatalf("cannot copy %s for the test: %v", pkiSetupKongJSON, err)
+		}
+	}
+
+	testResTomlFile := filepath.Join(testResourceDir, configTomlFile)
+	if _, err := copyFile(filepath.Join(resDir, configTomlFile), testResTomlFile); err != nil {
+		t.Fatalf("cannot copy %s for the test: %v", configTomlFile, err)
+	}
+
+	_ = setup.Init()
+
+	origScratchDir := pkiInitScratchDir
+	testScratchDir, tempDirErr := ioutil.TempDir(curDir, "scratch")
+	if tempDirErr != nil {
+		t.Fatalf("cannot create temporary scratch directory for the test: %v", tempDirErr)
+	}
+	pkiInitScratchDir = filepath.Base(testScratchDir)
+
+	origGeneratedDir := pkiInitGeneratedDir
+	testGeneratedDir, tempDirErr := ioutil.TempDir(curDir, "generated")
+	if tempDirErr != nil {
+		t.Fatalf("cannot create temporary generated directory for the test: %v", tempDirErr)
+	}
+	pkiInitGeneratedDir = filepath.Base(testGeneratedDir)
+
+	origEnvXdgRuntimeDir := os.Getenv(envXdgRuntimeDir)
+	// change it to the current working directory
+	os.Setenv(envXdgRuntimeDir, curDir)
+
+	origEnvPkiCache := os.Getenv(envPkiCache)
+	// use curDir/cache as the working directory for test
+	pkiCacheDir := filepath.Join(curDir, "cache")
+	os.Setenv(envPkiCache, pkiCacheDir)
+	_ = createDirectoryIfNotExists(pkiCacheDir)
+
+	origDeployDir := pkiInitDeployDir
+	tempDir, tempDirErr := ioutil.TempDir(curDir, "deploytest")
+	if tempDirErr != nil {
+		t.Fatalf("cannot create temporary scratch directory for the test: %v", tempDirErr)
+	}
+	pkiInitDeployDir = tempDir
+
+	return func(t *testing.T) {
```

Comment:
what's the purpose of this function taking `testing.T` as an argument? It doesn't seem to be used currently.

The `testing.T` that would normally be used would be the one that's passed here to the setup function so you could just use `t` here if you needed it, with go creating a closure around the `t` value for this function that's returned.

Reply:
good catch, can remove `testing.T`.  The only point to keep that is to be more symmetric interface signature with setup(t). but yeah, will remove it from this